File: benchmarks/marine_benchmark.py
#!/usr/bin/env python3
"""
Marine Algorithm Benchmark - Python Edition

Matches the Rust benchmark for fair comparison.
Let's see how Python's interpreted bytecode stacks up! 🐍
"""

import sys
import time
import json
import platform
import numpy as np
from typing import Dict, List

# Import our Marine implementation
sys.path.insert(0, '/aidata/Turbo-Whisper-Workspace')
from vocalis.marine.marine_localization import MarineAlgorithm


class BenchmarkResult:
    """Benchmark result container"""

    def __init__(
        self,
        duration_ms: float,
        samples_processed: int,
        samples_per_sec: float,
        realtime_factor: float,
        peak_count: int,
        max_salience: float
    ):
        self.duration_ms = duration_ms
        self.samples_processed = samples_processed
        self.samples_per_sec = samples_per_sec
        self.realtime_factor = realtime_factor
        self.peak_count = peak_count
        self.max_salience = max_salience

    def print_result(self, label: str):
        """Print benchmark result"""
        print("\n╔═══════════════════════════════════════════════════════════╗")
        print(f"║  {label:<57}║")
        print("╚═══════════════════════════════════════════════════════════╝")
        print(f"  Duration:        {self.duration_ms:.3f} ms")
        print(f"  Samples:         {self.samples_processed}")
        print(f"  Throughput:      {self.samples_per_sec / 1_000_000:.2f} Msamples/sec")
        print(f"  Real-time factor: {self.realtime_factor:.1f}x")
        print(f"  Peaks detected:  {self.peak_count}")
        print(f"  Max salience:    {self.max_salience:.2f}")
        print(f"  Memory/sample:   ~{np.dtype(np.float32).itemsize * 2} bytes")

    def to_dict(self) -> Dict:
        """Convert to dictionary for JSON export"""
        return {
            'duration_ms': float(self.duration_ms),
            'throughput_msamples': float(self.samples_per_sec / 1_000_000),
            'realtime_factor': float(self.realtime_factor),
            'peak_count': int(self.peak_count),
            'max_salience': float(self.max_salience)
        }


def generate_test_audio(sample_rate: int, duration_secs: float) -> np.ndarray:
    """
    Generate synthetic test audio

    Must match the Rust implementation exactly for fair comparison!
    """
    num_samples = int(sample_rate * duration_secs)
    t = np.linspace(0, duration_secs, num_samples, dtype=np.float32)

    # Quasi-periodic signal (speech-like)
    # Fundamental at 100Hz with harmonics
    fundamental = np.sin(2 * np.pi * 100 * t)
    harmonic2 = 0.5 * np.sin(2 * np.pi * 200 * t)
    harmonic3 = 0.25 * np.sin(2 * np.pi * 300 * t)

    audio = (fundamental + harmonic2 + harmonic3) * 0.3

    # Add slight noise
    noise = (np.random.rand(num_samples).astype(np.float32) - 0.5) * 0.05
    audio += noise

    return audio


def benchmark_marine(audio: np.ndarray, sample_rate: int, label: str) -> BenchmarkResult:
    """
    Benchmark Marine algorithm

    Matches Rust implementation methodology
    """
    marine = MarineAlgorithm(theta_c=0.015, alpha=0.15)

    # Warm-up run (not timed) - let JIT warm up if possible
    marine.process(audio, sample_rate)
    marine.reset()

    # Timed run
    start_time = time.perf_counter()
    result = marine.process(audio, sample_rate)
    end_time = time.perf_counter()

    duration = end_time - start_time
    duration_ms = duration * 1000.0
    samples_processed = len(audio)
    samples_per_sec = samples_processed / duration

    # Real-time factor
    audio_duration_secs = samples_processed / sample_rate
    realtime_factor = audio_duration_secs / duration

    return BenchmarkResult(
        duration_ms=duration_ms,
        samples_processed=samples_processed,
        samples_per_sec=samples_per_sec,
        realtime_factor=realtime_factor,
        peak_count=len(result['peaks']),
        max_salience=result['max_salience']
    )


def run_benchmark_suite():
    """Run comprehensive benchmark suite"""
    print("🐍 PYTHON MARINE ALGORITHM BENCHMARK")
    print("═══════════════════════════════════════════════════════════")
    print(f"Python:      {platform.python_version()}")
    print(f"NumPy:       {np.__version__}")
    print(f"Platform:    {platform.platform()}")
    print(f"Processor:   {platform.processor()}")
    print()

    sample_rate = 16000

    # Test 1: Short audio (1 second)
    print("Test 1: Short audio (1s @ 16kHz)")
    audio_1s = generate_test_audio(sample_rate, 1.0)
    result_1s = benchmark_marine(audio_1s, sample_rate, "1 Second Audio")
    result_1s.print_result("1 SECOND AUDIO")

    # Test 2: Medium audio (10 seconds)
    print("\nTest 2: Medium audio (10s @ 16kHz)")
    audio_10s = generate_test_audio(sample_rate, 10.0)
    result_10s = benchmark_marine(audio_10s, sample_rate, "10 Second Audio")
    result_10s.print_result("10 SECOND AUDIO")

    # Test 3: Long audio (60 seconds)
    print("\nTest 3: Long audio (60s @ 16kHz)")
    audio_60s = generate_test_audio(sample_rate, 60.0)
    result_60s = benchmark_marine(audio_60s, sample_rate, "60 Second Audio")
    result_60s.print_result("60 SECOND AUDIO")

    # Test 4: High sample rate (48kHz)
    print("\nTest 4: High sample rate (10s @ 48kHz)")
    audio_48k = generate_test_audio(48000, 10.0)
    result_48k = benchmark_marine(audio_48k, 48000, "48kHz Audio")
    result_48k.print_result("48kHz AUDIO")

    # Summary
    results = [result_1s, result_10s, result_60s, result_48k]
    avg_throughput = np.mean([r.samples_per_sec for r in results])
    avg_rt_factor = np.mean([r.realtime_factor for r in results])

    print("\n╔═══════════════════════════════════════════════════════════╗")
    print("║  BENCHMARK SUMMARY                                        ║")
    print("╚═══════════════════════════════════════════════════════════╝")
    print(f"  Average throughput: {avg_throughput / 1_000_000:.2f} Msamples/sec")
    print(f"  Average RT factor:  {avg_rt_factor:.1f}x")

    # Export results as JSON
    print("\n💾 Exporting results to python_benchmark_results.json")
    json_results = {
        'language': 'Python',
        'version': platform.python_version(),
        'numpy_version': np.__version__,
        'platform': platform.platform(),
        'tests': [
            {'name': '1s @ 16kHz', **result_1s.to_dict()},
            {'name': '10s @ 16kHz', **result_10s.to_dict()},
            {'name': '60s @ 16kHz', **result_60s.to_dict()},
            {'name': '10s @ 48kHz', **result_48k.to_dict()},
        ]
    }

    with open('python_benchmark_results.json', 'w') as f:
        json.dump(json_results, f, indent=2)

    print("✅ Benchmark complete!")

    return json_results


if __name__ == '__main__':
    # Try to set higher priority (best effort)
    try:
        import os
        os.nice(-10)
    except (PermissionError, AttributeError):
        pass

    run_benchmark_suite()

File: benchmarks/marine_benchmark.rs
//! Marine Algorithm Benchmark - Rust Edition
//!
//! Measures pure Rust Marine performance for comparison with Python
//! Demoscene-worthy optimization! Every cycle counts! 🚀

use std::time::{Duration, Instant};
use std::path::PathBuf;

// Import Marine from the marine-rust crate
// We'll link against the existing implementation
extern crate marine_algorithm;
use marine_algorithm::marine::{Marine, MarineAlgorithm};
use ndarray::Array1;

/// Benchmark result
#[derive(Debug)]
struct BenchmarkResult {
    duration_ms: f64,
    samples_processed: usize,
    samples_per_sec: f64,
    realtime_factor: f64,
    peak_count: usize,
    max_salience: f32,
}

impl BenchmarkResult {
    fn print(&self, label: &str) {
        println!("\n╔═══════════════════════════════════════════════════════════╗");
        println!("║  {}  ", label);
        println!("╚═══════════════════════════════════════════════════════════╝");
        println!("  Duration:        {:.3} ms", self.duration_ms);
        println!("  Samples:         {}", self.samples_processed);
        println!("  Throughput:      {:.2} Msamples/sec", self.samples_per_sec / 1_000_000.0);
        println!("  Real-time factor: {:.1}x", self.realtime_factor);
        println!("  Peaks detected:  {}", self.peak_count);
        println!("  Max salience:    {:.2}", self.max_salience);
        println!("  Memory/sample:   ~{} bytes", std::mem::size_of::<f32>() * 2); // Peak + score
    }
}

/// Generate synthetic test audio
fn generate_test_audio(sample_rate: u32, duration_secs: f32) -> Array1<f32> {
    let num_samples = (sample_rate as f32 * duration_secs) as usize;
    let mut audio = Array1::zeros(num_samples);

    let fs = sample_rate as f32;

    // Generate quasi-periodic signal similar to speech
    // Fundamental at 100Hz with harmonics
    for i in 0..num_samples {
        let t = i as f32 / fs;
        let fundamental = (2.0 * std::f32::consts::PI * 100.0 * t).sin();
        let harmonic2 = 0.5 * (2.0 * std::f32::consts::PI * 200.0 * t).sin();
        let harmonic3 = 0.25 * (2.0 * std::f32::consts::PI * 300.0 * t).sin();

        audio[i] = (fundamental + harmonic2 + harmonic3) * 0.3;

        // Add slight noise
        audio[i] += (rand::random::<f32>() - 0.5) * 0.05;
    }

    audio
}

/// Benchmark Marine algorithm
fn benchmark_marine(audio: &Array1<f32>, sample_rate: u32, label: &str) -> BenchmarkResult {
    let mut marine = MarineAlgorithm::new(0.015, 0.15);

    // Warm-up run (not timed)
    marine.process(audio, sample_rate);
    marine.reset();

    // Timed run
    let start = Instant::now();
    let result = marine.process(audio, sample_rate);
    let duration = start.elapsed();

    let duration_ms = duration.as_secs_f64() * 1000.0;
    let samples_processed = audio.len();
    let samples_per_sec = samples_processed as f64 / duration.as_secs_f64();

    // Real-time factor: how many times faster than real-time playback
    let audio_duration_secs = samples_processed as f64 / sample_rate as f64;
    let realtime_factor = audio_duration_secs / duration.as_secs_f64();

    BenchmarkResult {
        duration_ms,
        samples_processed,
        samples_per_sec,
        realtime_factor,
        peak_count: result.peaks.len(),
        max_salience: result.max_salience(),
    }
}

/// Run comprehensive benchmark suite
fn run_benchmark_suite() {
    println!("🦀 RUST MARINE ALGORITHM BENCHMARK");
    println!("═══════════════════════════════════════════════════════════");
    println!("Compiler: rustc {} (opt-level=3, lto=true)",
             rustc_version_runtime::version());
    println!("CPU: {}", std::env::var("PROCESSOR_IDENTIFIER")
             .unwrap_or_else(|_| "Unknown".to_string()));
    println!();

    let sample_rate = 16000;

    // Test 1: Short audio (1 second)
    println!("Test 1: Short audio (1s @ 16kHz)");
    let audio_1s = generate_test_audio(sample_rate, 1.0);
    let result_1s = benchmark_marine(&audio_1s, sample_rate, "1 Second Audio");
    result_1s.print("1 SECOND AUDIO");

    // Test 2: Medium audio (10 seconds)
    println!("\nTest 2: Medium audio (10s @ 16kHz)");
    let audio_10s = generate_test_audio(sample_rate, 10.0);
    let result_10s = benchmark_marine(&audio_10s, sample_rate, "10 Second Audio");
    result_10s.print("10 SECOND AUDIO");

    // Test 3: Long audio (60 seconds)
    println!("\nTest 3: Long audio (60s @ 16kHz)");
    let audio_60s = generate_test_audio(sample_rate, 60.0);
    let result_60s = benchmark_marine(&audio_60s, sample_rate, "60 Second Audio");
    result_60s.print("60 SECOND AUDIO");

    // Test 4: High sample rate (48kHz)
    println!("\nTest 4: High sample rate (10s @ 48kHz)");
    let audio_48k = generate_test_audio(48000, 10.0);
    let result_48k = benchmark_marine(&audio_48k, 48000, "48kHz Audio");
    result_48k.print("48kHz AUDIO");

    // Summary
    println!("\n╔═══════════════════════════════════════════════════════════╗");
    println!("║  BENCHMARK SUMMARY                                        ║");
    println!("╚═══════════════════════════════════════════════════════════╝");
    println!("  Average throughput: {:.2} Msamples/sec",
             (result_1s.samples_per_sec + result_10s.samples_per_sec +
              result_60s.samples_per_sec + result_48k.samples_per_sec) / 4.0 / 1_000_000.0);
    println!("  Average RT factor:  {:.1}x",
             (result_1s.realtime_factor + result_10s.realtime_factor +
              result_60s.realtime_factor + result_48k.realtime_factor) / 4.0);

    // Export results as JSON
    println!("\n💾 Exporting results to rust_benchmark_results.json");
    let json_results = serde_json::json!({
        "language": "Rust",
        "compiler": format!("rustc {}", rustc_version_runtime::version()),
        "optimization": "opt-level=3, lto=true, codegen-units=1",
        "tests": [
            {
                "name": "1s @ 16kHz",
                "duration_ms": result_1s.duration_ms,
                "throughput_msamples": result_1s.samples_per_sec / 1_000_000.0,
                "realtime_factor": result_1s.realtime_factor,
                "peak_count": result_1s.peak_count,
            },
            {
                "name": "10s @ 16kHz",
                "duration_ms": result_10s.duration_ms,
                "throughput_msamples": result_10s.samples_per_sec / 1_000_000.0,
                "realtime_factor": result_10s.realtime_factor,
                "peak_count": result_10s.peak_count,
            },
            {
                "name": "60s @ 16kHz",
                "duration_ms": result_60s.duration_ms,
                "throughput_msamples": result_60s.samples_per_sec / 1_000_000.0,
                "realtime_factor": result_60s.realtime_factor,
                "peak_count": result_60s.peak_count,
            },
            {
                "name": "10s @ 48kHz",
                "duration_ms": result_48k.duration_ms,
                "throughput_msamples": result_48k.samples_per_sec / 1_000_000.0,
                "realtime_factor": result_48k.realtime_factor,
                "peak_count": result_48k.peak_count,
            }
        ]
    });

    std::fs::write(
        "rust_benchmark_results.json",
        serde_json::to_string_pretty(&json_results).unwrap()
    ).expect("Failed to write results");

    println!("✅ Benchmark complete!");
}

fn main() {
    // Set high priority for accurate benchmarking
    #[cfg(target_os = "linux")]
    {
        use std::process::Command;
        let _ = Command::new("renice")
            .args(&["-n", "-10", "-p", &std::process::id().to_string()])
            .status();
    }

    run_benchmark_suite();
}
